velox/functions/sparksql/ArrayFunction.h

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <cmath>
+#include <type_traits>
+#include "velox/functions/Udf.h"
+
+namespace facebook::velox::functions::sparksql {
+template <typename TExecCtx, bool isMax>
+struct ArrayMinMaxFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(TExecCtx);
+
+  // Results refer to strings in the first argument.
+  static constexpr int32_t reuse_strings_from_arg = 0;
+
+  template <typename T>
+  void update(T& currentValue, const T& candidateValue) {
+    // NaN is greater than any non-NaN elements for double/float type.
+    if constexpr (std::is_floating_point_v<T>) {
+      if constexpr (isMax) {
+        if (std::isnan(candidateValue) ||
+            (!std::isnan(currentValue) && candidateValue > currentValue)) {
+          currentValue = candidateValue;
+        }
+      } else {
+        if (std::isnan(currentValue) ||
+            (!std::isnan(candidateValue) && candidateValue < currentValue)) {
+          currentValue = candidateValue;
+        }
+      }
+      return;
+    }
+
+    if constexpr (isMax) {
+      if (candidateValue > currentValue) {
+        currentValue = candidateValue;
+      }
+    } else {
+      if (candidateValue < currentValue) {
+        currentValue = candidateValue;
+      }
+    }
+  }
+
+  template <typename TReturn, typename TInput>
+  void assign(TReturn& out, const TInput& value) {
+    out = value;
+  }
+
+  void assign(out_type<Varchar>& out, const arg_type<Varchar>& value) {
+    out.setNoCopy(value);
+  }
+
+  template <typename TReturn, typename TInput>
```

Comment:
Yes, fix that.
By the way, this is copied from implementation under prestosql, maybe that implementation can be improved.

Reply:
``` cpp
 template <typename T>
  bool call(T& out, const ArrayVal<T>& array) {
```
Sorry, when I try to change `call()` signature above, It can't initialize the template for `Varchar`, any suggestions?

Reply:
Try

```
template <typename T> bool call(out_type<T>& out, const arg_type<Array<T>>& array) {
```

See https://facebookincubator.github.io/velox/develop/scalar-functions.html

Reply:
I have try with this, but I don't understand why can't infer `T` for `Varchar`
I also checked `out_type<Varchar> == StringWriter<false>` and `arg_type<Array<Varchar>> == ArrayView<true, Varchar>`

but got this error
```bash
../../velox/functions/sparksql/Register.cpp:47:18: error: no matching function for call to ‘facebook::velox::functions::sparksql::ArrayMinFunction<facebook::velox::exec::VectorExec>::call(facebook::velox::exec::StringWriter<false>&, const facebook::velox::exec::ArrayView<true, facebook::velox::Varchar>&)’
In file included from ../../velox/functions/sparksql/Register.cpp:26:
../.././velox/functions/sparksql/ArrayFunction.h:80:8: note: candidate: ‘template<class T> bool facebook::velox::functions::sparksql::ArrayMinMaxFunction<TExecCtx, isMax>::call(facebook::velox::functions::sparksql::ArrayMinMaxFunction<TExecCtx, isMax>::out_type<T>&, facebook::velox::functions::sparksql::ArrayMinMaxFunction<TExecCtx, isMax>::arg_type<facebook::velox::Array<TElement> >&) [with T = T; TExecCtx = facebook::velox::exec::VectorExec; bool isMax = false]’
   80 |   bool call(out_type<T>& out, const arg_type<Array<T>>& array) {
      |        ^~~~
../.././velox/functions/sparksql/ArrayFunction.h:80:8: note:   template argument deduction/substitution failed:
../../velox/functions/sparksql/Register.cpp:47:18: note:   couldn’t deduce template parameter ‘T’
   47 |   constexpr auto resolve(TArgs&&... args) const
      |                  ^~~~~~~
```

simplified the template utils below:
```cpp
template <typename T>
using exec_resolver =
    typename facebook::velox::exec::VectorExec::template resolver<T>;

struct CallMethodResolver {
  template <class C, typename... TArgs>
  constexpr auto resolve(TArgs&&... args) const
      -> decltype(std ::declval<C>().call(args...)) {
    return {};
  }
};

template <typename C, class TResolver, typename TRet, typename... TArgs>
struct HasMethod {
 private:
  template <typename CT>
  static constexpr auto check(CT * /*unused*/) -> typename std::is_same<
      decltype(std::declval<TResolver>().template resolve<CT>(
          std::declval<TArgs>()...)),
      TRet>::type {
    return {};
  }

  using type = decltype(check<C>(nullptr));

 public:
  static constexpr bool value = type::value;
};
```

```cpp
  using funcClass = ArrayMinFunction<exec::VectorExec>;

  using Type = Varchar;
  using exec_return_type = typename exec_resolver<Type>::out_type;
  using exec_arg_type = typename exec_resolver<Array<Type>>::in_type;

  static constexpr bool udf_has_call_return_bool = HasMethod<
      funcClass,
      CallMethodResolver,
      bool,
      exec_return_type,
      const exec_arg_type&>::value;
  static_assert(udf_has_call_return_bool, "call()");

```

Reply:
Got it. @laithsakka Laith, any chance you could help us understand what's going on here?